### ISSUE TITLE
no-use-before-define has false negative

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6699]
**What version of ESLint are you using?**

- v3.1.0

**What parser (default, Babel-ESLint, etc.) are you using?**

- default

**Please show your full configuration:**

- nothing

**What did you do? Please include the actual source code causing the issue.**

- Linted the following code:

```js
/*eslint no-undef:error, no-use-before-define:error*/
/*eslint-env es6*/

let a = {a: []}
for (let a of a.a) {
}
for (let a in a) {
}
```

**What did you expect to happen?**

`no-use-before-define` rule should warn references on `ForOfStatement#right` or `ForInStatement#right` (i.e. `a` of `of a.a` and `in a`)
Those are similar to `let a = a`.
`let a = a` is warned by `no-use-before-define` rule. (It would throw reference errors in runtime because TDZ.)

**What actually happened? Please include the actual, raw output from ESLint.**

![image](https://cloud.githubusercontent.com/assets/1937871/16902451/982d7e52-4c99-11e6-9237-4b5e975ff969.png)

No errors.

---

Refs: https://esdiscuss.org/topic/redefining-a-let-variable-inside-a-for-loop-scope-doesnt-work
